Series of small WCOSS2 updates - round 10

KateFriedman-NOAA opened the pull request:
**Description**

This PR includes some updates for the WCOSS2 port:

1. bug fix in `scripts/exgfs_atmos_gempak_meta.sh` which was found and fixed by @WeiWei-NCO
2. resource updates for several GSI/EnKF jobs based on feedback from @RussTreadon-NOAA and @MichaelLueken-NOAA 
3. cleans up nco.static resource configs to support only operations on WCOSS2; the emc.dyn resource configs support other tier 1 platforms
4. updates the EMC tag name in the release notes to the new tag that will be cut after this PR goes in (`EMC-v16.2.0.3`)

Refs: #398, #399

**Type of change**

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] Regular port updates - resources

**How Has This Been Tested?**

Both @WeiWei-NCO and myself tested the included updates on Dogwood. We confirmed good timings on adjusted jobs.

KateFriedman-NOAA committed:
Update to correct infinite loop in gempak script

- NCO/Wei found an infinite loop in the exgfs_atmos_gempak_meta.sh
script during para testing
- Provided and tested a fix, confirmed fix with Boi

Refs: #399

KateFriedman-NOAA committed:
Update several GSI/EnKF job resources

- Russ Treadon provided resource adjustments to several GSI/EnKF jobs to
get them closer to their WCOSS1 ops timings and use fewer cores
- The enkfgdas_select_obs (gdaseobs), enkfgdas_update (gdaseupd),
gdas_atmos_analysis_diag (gdasanaldiag), and gdas_atmos_analysis
(gdasanal) jobs have been updated
- The updated resources have been incorporated into the relevant ecf
script PBS directives and resource configs

Refs: #399

RussTreadon-NOAA commented on parm/config/config.resources.emc.dyn:
```diff
@@ -160,7 +160,7 @@ elif [ $step = "analcalc" ]; then
 
     export wtime_analdiag="00:10:00"
-    export npe_analdiag=72
+    export npe_analdiag=115		# Should be at least twice npe_ediag
```
Setting this to twice npe_ediag is sufficient.   Currently npe_ediag=48.   Therefore, set npe_analdiag=96..   If concerned that this is too low, use the operational WCOSS_D values of npe_ediag=56 and npe_analdiag=112.  Similar comment applies to `config.resources.nco.static`

KateFriedman-NOAA replied:
Noted, thanks @RussTreadon-NOAA !

RussTreadon-NOAA commented on parm/config/config.resources.emc.dyn:
```diff
@@ -338,7 +339,8 @@ elif [ $step = "eupd" ]; then
     fi
-    export npe_node_eupd=10
+    export npe_node_eupd=$(echo "$npe_node_max / $nth_eupd" | bc)
+    if [[ "$machine" = "WCOSS2" ]]; then export npe_node_eupd=9; fi
```
This line is not necessary for WCOSS2.   128 / 14 = 9.142785... or 9 with integer math.   Similar comment applies to `config.resources.nco.static`

KateFriedman-NOAA replied:
@RussTreadon-NOAA You are correct, I have removed this line from both config.resource* files:
`if [[ "$machine" = "WCOSS2" ]]; then export npe_node_eupd=9; fi`
Generally, when the setup script math doesn't evenly divide (e.g. 128/14=9.1....) I tend to make sure the settings are forced so rocoto doesn't round up. For this situation though it rounds down so it's fine to not force it.

KateFriedman-NOAA committed:
Remove npe_node_eupd=9 setting on WCOSS2

- not needed since setup script math results in 9ppn for eupd when using
14 threads

Refs: #399

RussTreadon-NOAA commented on ecf/scripts/gdas/atmos/analysis/jgdas_atmos_analysis_diag.ecf:
```diff
@@ -4,7 +4,7 @@
 #PBS -A %PROJ%-%PROJENVIR%
 #PBS -l walltime=00:10:00
-#PBS -l place=vscatter,select=1:mpiprocs=72:ompthreads=1:ncpus=72:mem=48GB
+#PBS -l place=vscatter,select=1:mpiprocs=115:ompthreads=1:ncpus=115:mem=48GB
```
Please set mpiprocs=96 to that it is twice mpiprocs in `jenkfgdas_diag.ecf`.

KateFriedman-NOAA replied:
Changing back to 96.

RussTreadon-NOAA commented on parm/config/config.resources.emc.dyn:
```diff
@@ -160,7 +160,7 @@ elif [ $step = "analcalc" ]; then
 
     export wtime_analdiag="00:10:00"
-    export npe_analdiag=72
+    export npe_analdiag=115		# Should be at least twice npe_ediag
```
Please set npe_analdiag=96 so that it is twice npe_ediag.  Same comment applies to `config.resources.nco.static`.

KateFriedman-NOAA replied:
Will change this back to 96. Thanks @RussTreadon-NOAA !

WalterKolczynski-NOAA commented on ecf/scripts/enkfgdas/analysis/create/jenkfgdas_update.ecf:
```diff
@@ -4,7 +4,7 @@
 #PBS -A %PROJ%-%PROJENVIR%
 #PBS -l walltime=00:30:00
-#PBS -l place=vscatter,select=32:mpiprocs=10:ompthreads=12:ncpus=120
+#PBS -l place=vscatter,select=35:mpiprocs=9:ompthreads=14:ncpus=126
```
Any reason you've set a node count that doesn't divide evenly?

RussTreadon-NOAA replied:
The suggested settings were driven by the requirement to obtain a WCOSS2 job run time comparable to WCOSS_D.   Other combinations may achieve a similar end.    Exploring this parameter space takes time.

WalterKolczynski-NOAA replied:
Just found it weird that different nodes were getting different numbers of tasks.

WalterKolczynski-NOAA commented on ecf/scripts/gdas/atmos/analysis/jgdas_atmos_analysis.ecf:
```diff
@@ -4,7 +4,7 @@
 #PBS -A %PROJ%-%PROJENVIR%
 #PBS -l walltime=00:40:00
-#PBS -l place=vscatter,select=55:mpiprocs=15:ompthreads=8:ncpus=120
+#PBS -l place=vscatter,select=52:mpiprocs=15:ompthreads=8:ncpus=120
```
Same node count question here.

RussTreadon-NOAA replied:
Same response as above.

WalterKolczynski-NOAA commented on parm/config/config.resources.emc.dyn:
```diff
@@ -302,9 +302,10 @@ elif [ $step = "eobs" -o $step = "eomg" ]; then
+    export nth_eobs=3
     if [[ "$machine" = "WCOSS_DELL_P3" ]]; then export nth_eobs=7; fi
     export npe_node_eobs=$(echo "$npe_node_max / $nth_eobs" | bc)
+    if [[ "$machine" = "WCOSS2" ]]; then export npe_node_eobs=40; fi
```
Why is this override here? The math works out to 40 anyway, doesn't it?

KateFriedman-NOAA replied:
If I don't force it to be 40 the setup script math sets it to 42 (128/3=42.6). Here is the xml diff if I remove that line and rerun setup_workflow.py:
```
Kate.Friedman@dlogin03:/lfs/h2/emc/global/noscrub/Kate.Friedman/expdir/testcyc768o> diff testcyc768o.xml testcyc768o.xml_23
13c13
<               This workflow was automatically generated at 2022-01-31 20:10:45.310360
---
>               This workflow was automatically generated at 2022-01-31 16:22:16.683725
143c143
<       <!ENTITY RESOURCES_EOBS_GDAS "<nodes>12:ppn=42:tpp=3</nodes>">
---
>       <!ENTITY RESOURCES_EOBS_GDAS "<nodes>12:ppn=40:tpp=3</nodes>">
```

RussTreadon-NOAA replied:
`npe_node_max` is 128 on WCOSS2.   128 / 3 = 42.66667.   This doesn't round to 40.

WalterKolczynski-NOAA replied:
Ah, I thought it was 120. Sorry.

WalterKolczynski-NOAA commented on parm/config/config.resources.nco.static:
```diff
@@ -302,9 +302,10 @@ elif [ $step = "eobs" -o $step = "eomg" ]; then
+    export nth_eobs=3
     if [[ "$machine" = "WCOSS_DELL_P3" ]]; then export nth_eobs=7; fi
     export npe_node_eobs=$(echo "$npe_node_max / $nth_eobs" | bc)
+    if [[ "$machine" = "WCOSS2" ]]; then export npe_node_eobs=40; fi
```
As before

RussTreadon-NOAA replied:
Same reply as before.

WalterKolczynski-NOAA reviewed (changes requested):
Some questions about uneven node counts for @RussTreadon-NOAA

KateFriedman-NOAA committed:
Change npe_analdiag to 96

Refs: #399

KateFriedman-NOAA committed:
Update EMC tag name in release notes

- the new EMC hand-off tag is EMC-v16.2.0.3

Refs: #399

KateFriedman-NOAA committed:
Remove non-WCOSS2 references in nco.static configs

- clean up config.resources.nco.static and config.fv3.nco.static to
remove non-WCOSS2 and non-ops refererences
- both nco.static resource configs now only support operations on WCOSS2
and preserve the operational resource settings
- the emc.dyn resource configs will be modified further to support other
tier 1 platforms

Refs: #399

WalterKolczynski-NOAA commented on parm/config/config.fv3.nco.static:
```diff
@@ -19,21 +19,7 @@ case_in=$1
+export npe_node_max=128
 
 # (Standard) Model resolution dependent variables
 case $case_in in
```
Most of this can go too.

KateFriedman-NOAA replied:
I'll leave the other CASE blocks in config.fv3.nco.static for now, just in case. Will consider removing in the develop branch version. Thanks!

WalterKolczynski-NOAA reviewed (changes requested):
You could theoretically replace all of the math with constants, but that might be too much work for now.
